### ISSUE TITLE
Negotiate-module fails on new version of PHP krb5-module

### ISSUE THREAD

**Issue by tvdijen** [simplesamlphp/simplesamlphp#475]
Recently version 1.1.0 of the PHP krb5-module was released (https://pecl.php.net/package/krb5).
After installing it, Kerberos-authentication stopped working. No error-messages, just a blank IE-page telling me 'page cannot be displayed'.

Further examination of the log-files show me that the request is coming in:
`192.168.1.208 - - [16/Sep/2016:10:55:12 +0200] "GET /saml2/idp/SSOService.php?SAMLRequest=[...] HTTP/1.1" 401 1035`

And that the Negotiate-module kicks in:
```
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Session: 'KERBEROS' not valid because we are not authenticated.
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Negotiate - SP disable flag not found
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Negotiate - authenticate(): looking for Negotate
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Negotiate - authenticate(): Sending Negotiate.
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Negotiate - fallback: LDAP
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Saved state: '_2c8233ad3d24a9da64703f33b8da476d15546c60ed:https://idp.moo-archive.nl/saml2/idp/SSOService.php?spentityid=https%3A%2F%2Fdemo.moo-arc$
Sep 16 10:55:13 simplesamlphp INFO [fa728ff0e2] SAML2.0 - IdP.SSOService: Accessing SAML 2.0 IdP endpoint SSOService
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Received message:
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] <samlp:AuthnRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="_d7d7be3bf515b06e8c6f$
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2]   <saml:Issuer>https://demo.moo-archive.nl/simplesaml/module.php/saml/sp/metadata.php/default-sp</saml:Issuer>
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2]   <samlp:NameIDPolicy Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient" AllowCreate="true"/>
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] </samlp:AuthnRequest>
Sep 16 10:55:13 simplesamlphp INFO [fa728ff0e2] SAML2.0 - IdP.SSOService: incoming authentication request: 'https://demo.moo-archive.nl/simplesaml/module.php/saml/sp/metadata.php/default-sp'
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Session: 'KERBEROS' not valid because we are not authenticated.
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Negotiate - SP disable flag not found
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Negotiate - authenticate(): looking for Negotate
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Negotiate - authenticate(): Negotate found
Sep 16 10:55:13 simplesamlphp DEBUG [fa728ff0e2] Library - LDAP __construct(): Setup LDAP with host='ldaps://dc.moo-archive.nl', tls=true, debug=true, timeout=3, referrals=false

```
And then it just turns silent everywhere in the log-files...

I have been able to reproduce this on both RedHat 6 and CentOs 7 on different AD-domains.
The fact that the example-code in the module's [docs](https://github.com/simplesamlphp/simplesamlphp/blob/master/modules/negotiate/docs/negotiate.md) _does_ work on both versions of the PHP-module, leads me to believe that somewhere in the SSP-module we are incorrectly using the PHP-module.

This obviously needs more investigation, and I'd be glad to help with it, but I'm a bit clueless at this point on how to approach this. It could as well be an issue with the PHP-module but since no bugs were reported since July, that's less likely.. I've examined their code-changes and I really can't find any clues to what might go wrong.

Any help / thoughts appreciated!

**Comment by thijskh** [comment]
I don't know anything about the specific PECL module, just some general thoughts on how to approach it. From the debug log it seems that the last successful thing that happened is [122](https://github.com/simplesamlphp/simplesamlphp/blob/master/modules/negotiate/lib/Auth/Source/Negotiate.php#L122). So it seems that [140](https://github.com/simplesamlphp/simplesamlphp/blob/master/modules/negotiate/lib/Auth/Source/Negotiate.php#L140) might be failing somehow. I'd put debug statements directly around that call to know for sure that this is indeed the failing part (and if not, look further of course).

If this is failing it's worthwhile to see if you can somehow determine why instantiating this object fails and/or how this is supposed to be done. Or try to create some standalone code that just tries to instantiate the object, and see if you can reproduce it in a minimal PHP script and under which conditions.

**Comment by tvdijen** [comment]
I went with the debug-lines and found out that doAuthentication on line [143](https://github.com/simplesamlphp/simplesamlphp/blob/master/modules/negotiate/lib/Auth/Source/Negotiate.php#L143) is the culprit here. Nothing happens after, no error is being catched in the catch-clause.

I also noticed that the earlier mentioned example code is failing on the same line once in a while..
The behaviour is quite random.. Sometimes it's working fine, but more often it fails..

I think it's safe to assume the problem is in the krb-module and not with the SSP-module.
I'll file a bug-report there.
